### PR TITLE
[Spec] Rainbow Bridge

### PR DESCRIPTION
**Pull request by amgando**
adding draft for review

### CODE REVIEW

File: specs/RainbowBridge/README.md [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+### Use Cases
+
+Please refer to the diagram above where (A), (B) and (C) appear to the left and (1), (2) and (3) appear the right.
+
+#### Use Case A-B-C-B-A 
+
+*NEAR smart contract must verify an event was recorded on the Ethereum blockchain*
+
+- A contract running on the *NEAR blockchain* requires verifiable proof that an event has been recorded on the *Ethereum blockchain*.  
+- The contract `(A)` invokes a method on the `EthEventVerifier` `(B)`, passing in the appropriate parameters to represent the event
```
Comment (k06a) [inline specs/RainbowBridge/README.md]:
```suggestion
- The contract `(A)` invokes a method on the `EthEventVerifier` `(B)`, passing in the appropriate parameters to represent the event and corresponding inclusion proof
```

File: specs/RainbowBridge/README.md [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+- `EthEventVerifier` `(B)` unpacks the Ethereum event and extracts the related block header which it passes on to the `EthClient` `(C)`
```
Comment (k06a) [inline specs/RainbowBridge/README.md]:
```suggestion
- `EthEventVerifier` `(B)` verifies the Ethereum event inclusion into the corresponding block header which exists in the `EthClient` `(C)`
```

File: specs/RainbowBridge/README.md [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+- `EthClient` `(C)` returns a boolean indicating whether the block header was indeed recorded on the Ethereum blockchain
```
Comment (k06a) [inline specs/RainbowBridge/README.md]:
```suggestion
- `EthClient` `(C)` returns a boolean indicating whether the block hash was indeed recorded on the Ethereum blockchain
```

File: specs/RainbowBridge/README.md [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+- `EthEventVerifier` `(B)` passes this result on to the contract `(A)` as proof or rejection of the event as valid
```
Comment (k06a) [inline specs/RainbowBridge/README.md]: Rejection?

File: specs/RainbowBridge/README.md [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+
+#### Use Case 1-2-3-2-1
+
+*Ethereum smart contract must verify a transaction result was recorded on the NEAR blockchain*
+
+- A contract running on the *Ethereum blockchain* requires verifiable proof that a transaction result has been recorded on the *NEAR blockchain*.  
+- The contract `(1)` invokes a method on the `NearTxResultVerifier` `(2)`, passing in the appropriate parameters to represent the transaction result
```
Comment (k06a) [inline specs/RainbowBridge/README.md]:
```suggestion
- The contract `(1)` invokes a method on the `NearTxResultVerifier` `(2)`, passing in the appropriate parameters to represent the transaction result and inclusion proof
```

File: specs/RainbowBridge/README.md [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+- `NearTxResultVerifier` `(2)` unpacks the Near transaction result and extracts the related block header which it passes on to the `NearClient` `(3)`
```
Comment (k06a) [inline specs/RainbowBridge/README.md]:
```suggestion
- `NearTxResultVerifier` `(2)` verifies the Near transaction belongs to the related block header which is stored in the `NearClient` `(3)`
```

File: specs/RainbowBridge/README.md [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+- `NearClient` `(3)` returns a boolean indicating whether the block header was indeed recorded on the NEAR blockchain
```
Comment (k06a) [inline specs/RainbowBridge/README.md]:
```suggestion
- `NearClient` `(3)` returns a boolean indicating whether the block hash was indeed recorded on the NEAR blockchain
```

File: specs/RainbowBridge/README.md [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+- `NearTxResultVerifier` `(2)` passes this result on to the contract `(1)` as proof or rejection of the transaction result as valid
```
Comment (k06a) [inline specs/RainbowBridge/README.md]: Rejection?

File: specs/RainbowBridge/README.md [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+
+
+### Software
+
+**Core Components**
+
+Two smart contracts, `(C)` and `(3)` establish the baseline source of truth for blockchain activity on the "other" network.
+
+- (C) `NearClient`- *smart contract hosted on **Ethereum** network*  \
+   This contract is a NEAR light client that receives NEAR block headers.  It verifies and stores block hashes only.
```
Comment (k06a) [inline specs/RainbowBridge/README.md]:
```suggestion
   This contract is a NEAR light client that receives NEAR block headers. It verifies the NEAR chain of blocks following Near Light Client Specification and stores block hashes only.
```

File: specs/RainbowBridge/README.md [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+
+- (3) `EthClient` – *smart contract hosted on the **NEAR** network*  \
+   This contract is an Ethereum light client that receives Ethereum block headers.  It verifies ethash and longest chain rule and stores block hashes only.
```
Comment (k06a) [inline specs/RainbowBridge/README.md]:
```suggestion
   This contract is an Ethereum light client that receives Ethereum block headers. It verifies ethash and longest chain rule and stores block hashes only.
```

File: specs/RainbowBridge/README.md [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+
+**Reference Implementations**
+
+Two smart contracts demonstrate what is possible using the Rainbow Bridge.  These contract are intended to be used as **reference material** for developing other contracts about various blockchain platforms. 
+
+*A note about NEAR and Ethereum blockchains*
+
+- At its core, NEAR is a sharded transaction processing engine which records *transaction results*.  These *transaction results* are a verifiable source of truth about the NEAR blockchain.
```
Comment (k06a) [inline specs/RainbowBridge/README.md]:
```suggestion
- At its core, NEAR is a sharded transaction processing engine that records *transaction results*. These *transaction results* are a verifiable source of truth about the NEAR blockchain.
```

File: specs/RainbowBridge/README.md [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+- At its core, Ethereum is single threaded computer which, at various points in time, records *events*.  These *events* are a verifiable source of truth about the Ethereum blockchain.
```
Comment (k06a) [inline specs/RainbowBridge/README.md]:
```suggestion
- At its core, Ethereum is a single-threaded computer that, at various points in time, records *events*. These *events* are a verifiable source of truth about the Ethereum blockchain.
```

File: specs/RainbowBridge/README.md [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+
+**Supporting Scripts**
+
+- `NearRelay` - an arbitrary, trustless actor that forwards NEAR block headers to the `NearClient` smart contract which is hosted on the Ethereum network.
+- `EthRelay` - an arbitrary, trustless actor that forwards Ethereum block headers to the `EthClient` smart contract which is hosted on the NEAR network.
+
+
+## Setting up the Rainbow Bridge
+
+1. clone the repo `git clone git@github.com:nearprotocol/near-bridge.git`
```
Comment (k06a) [inline specs/RainbowBridge/README.md]:
```suggestion
1. clone the repo `git clone --recurse-submodules git@github.com:nearprotocol/near-bridge.git`
```

Comment (k06a) [reply specs/RainbowBridge/README.md]: We have Git Submodule in `ethrelay/ethashproof` dir. We also need to call it's `build.sh` script to make EthRelay work.

File: specs/RainbowBridge/README.md [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+#[near_bindgen]
+impl EthBridge {
+    pub fn init(validate_ethash: bool, dags_start_epoch: u64, dags_merkle_roots: Vec<H128>) -> Self {}
+    pub fn apply_merkle_proof(&self, index: u64) -> H128 {}
+    pub fn initialized(&self) -> bool {}
+    pub fn last_block_number(&self) -> u64 {}
+    pub fn dag_merkle_root(&self, epoch: u64) -> H128 {}
+    pub fn block_hash(&self, index: u64) -> Option<H256> {}
+    pub fn block_hash_safe(&self, index: u64) -> Option<H256> {}
+    pub fn add_block_header(&mut self, block_header: Vec<u8>, dag_nodes: Vec<DoubleNodeWithMerkleProof>) {}
```
Comment (k06a) [inline specs/RainbowBridge/README.md]:
```suggestion
    // EthRelay-specific API methods:
    pub fn last_block_number(&self) -> u64;
    pub fn add_block_header(&mut self, block_header: Vec<u8>, dag_nodes: Vec<DoubleNodeWithMerkleProof>);

    // Public smart contracts API methods:
    pub fn block_hash(&self, index: u64) -> Option<H256>;
    pub fn block_hash_safe(&self, index: u64) -> Option<H256>;
```

File: specs/RainbowBridge/README.md [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+
+**Block header**
+
+*Please refer to [nearcore block primitives](https://github.com/nearprotocol/nearcore/blob/master/core/primitives/src/block.rs) for details*
+
+```
+struct ValidatorStake
+    account_id              AccountId               Account that stakes money.
+    public_key              PublicKey               Public key of the proposed validator.
+    amount                  Balance                 Stake / weight of the validator.
```
Comment (k06a) [inline specs/RainbowBridge/README.md]: Maybe better to keep it Java/C style, like:
```java
struct ValidatorStake {
      AccountId  account_id; // Account that stakes money.
      PublicKey public_key;  // Public key of the proposed validator.
      Balance amount;
}
```

File: specs/RainbowBridge/README.md (hunk 4 above)
Comment (amgando) [reply specs/RainbowBridge/README.md]: i mean that the event is valid or invalid, proof it is valid or rejection of the claim that it is valid

File: specs/RainbowBridge/README.md (hunk 8 above)
Comment (amgando) [reply specs/RainbowBridge/README.md]: same as above

File: specs/RainbowBridge/README.md (hunk 15 above)
Comment (amgando) [reply specs/RainbowBridge/README.md]: ✅

File: specs/RainbowBridge/README.md (hunk 13 above)
Comment (amgando) [reply specs/RainbowBridge/README.md]: ah, that's where that went!  thank you

File: specs/RainbowBridge/README.md (hunk 14 above)
Comment (amgando) [reply specs/RainbowBridge/README.md]: ✅

File: specs/RainbowBridge/README.md (hunk 12 above)
Comment (amgando) [reply specs/RainbowBridge/README.md]: ✅

File: specs/RainbowBridge/README.md (hunk 11 above)
Comment (amgando) [reply specs/RainbowBridge/README.md]: ✅

File: specs/RainbowBridge/README.md (hunk 9 above)
Comment (amgando) [reply specs/RainbowBridge/README.md]: ✅

File: specs/RainbowBridge/README.md (hunk 10 above)
Comment (amgando) [reply specs/RainbowBridge/README.md]: ✅

**Comment by ilblackdragon** [comment]
IMO This still missing the first things that should have been written in spec - which is exactly how it suppose to work, what’s the security parameters and limitations in README.

Here is an example:
https://github.com/cosmos/ics/blob/master/ibc/2_IBC_ARCHITECTURE.md

E.g I would expect opening README.md something like this:

 - Definition - This is what this thing is, what is the role it's playing, what can be done with it overalll
 - Scope - what is the scope of this spec, what things are included in this design vs what is separate standards or defined by developers
 - Data Flow - describe overall data flow (you have great diagram, but unless you already know what it does it's unclear without step by step for general case)  **This is the most important part IMO**
 - Security - explain what are security parameters, how much value can you load this with, etc. **This is second most important part**
 - Limitations - what are things you can not use this for.
 - Design Principles for {Verifiers, User contracts}  - what are APIs, etc.
 - Use Cases - list the use cases that this supports.
 - Separate docs can be for defining specific verifiers 

I'm not sure we need all the data structs listed that you have right now.
For the specific NEAR verifier - you can just link them to the other places in the spec where they should be.

**Comment by frol** [comment]
This PR seems to be severely outdated by now, so I am closing it. Please, feel free to submit a new NEP following the new NEP-001 guidelines.

**Pull request closed without merging**